Fix #960: S3 Feed Export throws boto error

jazzthief opened the pull request:
Fixes #960, closes #5735 by switching from `botocore` to `boto3` upload method, which supports multipart upload.

- Document the need to install boto3 for S3 support, rather than botocore.
- Implement code that uploads using boto3’s method instead of botocore to upload files. It seems the interfaces are similar.
- If boto3 is not installed, but botocore is, fall back to the current implementation, and log a deprecation warning.

jazzthief commented on docs/topics/feed-exports.rst:
```diff
@@ -101,12 +101,12 @@ The storages backends supported out of the box are:
 
 Some storage backends may be unavailable if the required external libraries are
-not available. For example, the S3 backend is only available if the botocore_
+not available. For example, the S3 backend is only available if at least the botocore_
```
Not quite sure how to properly formulate the library requirements in the docs. Strictly speaking, `S3FeedStorage` won't work only there is no botocore, so unless we want to deprecate botocore altogether - we should probably mention the fallback thing in more detail.

jazzthief commented on docs/topics/feed-exports.rst:
```diff
@@ -193,7 +193,7 @@ The feeds are stored on `Amazon S3`_.
     -   ``s3://aws_key:aws_secret@mybucket/path/to/export.csv``
 
--   Required external libraries: `botocore`_ >= 1.4.87
+-   Required external libraries: `boto3`_ >= 1.26.70, will fall back to botocore_ if unavailable
```
No particular reasoning behind the version, it's just what I have installed

jazzthief commented:
This works fine with `put_object()`. Switching to `upload_fileobj()` fails a number of tests - they probably need to be modified to support it.
Currently I'm thinking that existing tests should remain and cover only botocore's functionality, and I should add modified tests for boto3's functionality, possibly splitting them into single and multipart upload cases

jazzthief commented:
Fixed the call to botocore `get_session()` - now botocore functionality works as it should

jazzthief commented:
According to [this issue](https://github.com/boto/botocore/issues/974#issuecomment-229821575), `upload_file()` cannot be stubbed directly with botocore's `Stubber` that is currently used in the tests. I am guessing I can trace which exact methods `upload_file()` will call under the hood and stub them - or try a library like [placebo](https://github.com/garnaat/placebo) as botocore's maintainer suggested in the issue.
@Gallaecio, could you please have a look?

Gallaecio commented:
I will have a look, though it might take me some days :sweat:

Gallaecio reviewed (commented):
The change look good, I’ll try running the tests locally and see if I can figure out how to address the issues you mention.

jazzthief commented:
> The change look good, I’ll try running the tests locally and see if I can figure out how to address the issues you mention.

Got you, I will address your comments and wait for any thoughts you might have on tests

Gallaecio commented:
@jazzthief I see the `extra-deps-pinned` environment I added is causing issues.

If you can look into it while you work on my earlier feedback, that would be great. But I can take a look myself as well if you prefer.

jazzthief commented:
> If you can look into it while you work on my earlier feedback, that would be great. But I can take a look myself as well if you prefer.

I'll have a look. One question, though: I see `is_boto3_available()` used in tests - should I still switch everything to an `IS_BOTO3_AVAILABLE` variable inside `feedexport.py` or do we leave it like that?

Gallaecio commented:
I think it may be worth it switching to a variable, I just did not want to make it as part of my changes, and instead focus on the test part.

jazzthief commented:
@Gallaecio I addressed your comments.

Regarding `extra-deps-pinned`: from what I understood from boto3 changelog, the API for upload was completely implemented in 1.1.4, however it seems something was changed later, so I've put 1.20.0 into `extra-deps-pinned` config as well as the docs. This passed the tests locally.

Gallaecio reviewed (approved):
Seems to me like this is ready for review. Great job!

jazzthief commented:
> Seems to me like this is ready for review. Great job!

Thanks for your help! Do you think we should create an issue on reworking boto3 tests?

Gallaecio commented:
> reworking boto3 tests

Could you elaborate?

As far as I am concerned, test-wise we are covered here. While something in line with Stub would be great, I think mocking on our side just to make sure that we are mapping parameters correctly to what boto3 expects is a good approach, and not that different from stubbing.

jazzthief commented:
> I think mocking on our side just to make sure that we are mapping parameters correctly to what boto3 expects is a good approach, and not that different from stubbing.

That's what I meant. Got you, great then.

wRAR commented:
Thanks!
